### ISSUE TITLE
Feature request：Optimize pre-render pages in different language

### ISSUE THREAD

**Issue by zeyuyuan** [brillout/vite-plugin-ssr#136]
Expected behavior:
Avoid add .server.js/.route.js for every page to pre-render pages in different languages.  

demo is here: https://github.com/zeyuyuan/frontend-templates

**Comment by brillout** [comment]
```js
// _default.page.route.js

export { onBeforeRoute }

function onBeforeRoute(pageContext) {
  let { url } = pageContext

  const urlParts = url.split('/')
  const language = urlParts.splice(1, 1)
  url = urlParts.join('/')

  return {
    pageContext: {
      // `pageContext.language` will be available throughout our app.
      // And with https://vite-plugin-ssr.com/pageContext-anywhere we can access it
      // in all of our React/Vue components.
      language,
      // This will overwrite the original URL
      url,
    }
  }
}
```

Thoughts?

**Comment by zeyuyuan** [comment]
> ```js
> // _default.page.route.js
> 
> export { onBeforeRoute }
> 
> function onBeforeRoute(pageContext) {
>   let { url } = pageContext
> 
>   const urlParts = url.split('/')
>   const language = urlParts.splice(1, 1)
>   url = urlParts.join('/')
> 
>   return {
>     pageContext: {
>       // `pageContext.language` will be available throughout our app.
>       // And with https://vite-plugin-ssr.com/pageContext-anywhere we can access it
>       // in all of our React/Vue components.
>       language,
>       // This will overwrite the original URL
>       url,
>     }
>   }
> }
> ```
> 
> Thoughts?

Is this possible to work with pre-render, how does it recognize which language to pre-render in build-time?

**Comment by brillout** [comment]
Yes the `onBeforeHook` is always called upon routing, also while pre-rendering.

Would `onBeforeHook` work for you?

**Comment by zeyuyuan** [comment]
> Yes the `onBeforeHook` is always called upon routing, also while pre-rendering.
> 
> Would `onBeforeHook` work for you?

And i can move export { prerender } to _default.server.js? Ok,it should work for me.

**Comment by brillout** [comment]
>  export { prerender } to _default.server.js

Why do you want this? Why not export it in the page's `.page.server.js`?

**Comment by zeyuyuan** [comment]
> > export { prerender } to _default.server.js
> 
> Why do you want this? Why not export it in the page's `.page.server.js`?

I don’t want to add `.page.server.js` for  every page

**Comment by brillout** [comment]
How about using the `addPageContext()` hook in `_default.page.server.js` instead of using the `prerender()` hook for each page?

**Comment by zeyuyuan** [comment]
> How about using the `addPageContext()` hook in `_default.page.server.js` instead of using the `prerender()` hook for each page?

How to do it exactly?

**Comment by brillout** [comment]
Ok I see the problem. Let me think of a solution.

**Comment by brillout** [comment]
How about:

```js
// _default.page.server.js

export { onBeforePrerender }

function onBeforePrerender(pageContext) {
  const prerenderedPageContexts = []
  pageContext._prerenderedPageContexts.forEach(pageContext => {
    ['en-US', 'de-DE', 'fr-FR'].forEach(language => {
      prerenderedPageContexts.push({
        url: `/${language}${pageContext.url},
        language,
        ...pageContext
      })
    })
  })
  pageContext._prerenderedPageContexts = prerenderedPageContexts
}
```

Combined with `onBeforeRoute` I believe it will work.

Would that work for you? Thoughts?

**Comment by zeyuyuan** [comment]
> How about:
> 
> ```js
> // _default.page.server.js
> 
> export { onBeforePrerender }
> 
> function onBeforePrerender(pageContext) {
>   const prerenderedPageContexts = []
>   pageContext._prerenderedPageContexts.forEach(pageContext => {
>     ['en-US', 'de-DE', 'fr-FR'].forEach(language => {
>       prerenderedPageContexts.push({
>         url: `/${language}${pageContext.url},
>         language,
>         ...pageContext
>       })
>     })
>   })
>   pageContext._prerenderedPageContexts = prerenderedPageContexts
> }
> ```
> 
> Combined with `onBeforeRoute` I believe it will work.
> 
> Would that work for you? Thoughts?

Nice

**Comment by brillout** [comment]
Still thinking about some i18n related things.

ETA this/next week. Ok for you?

**Comment by zeyuyuan** [comment]
> Still thinking about some i18n related things.
> 
> ETA this/next week. Ok for you?

Yes, of course.

**Comment by brillout** [comment]
Everything's ready, I'll release the new `vite-plugin-ssr` version tomorrow.

You can get a sneak peak at https://vite-plugin-ssr.com/i18n#with-pre-rendering and https://github.com/brillout/vite-plugin-ssr/tree/master/examples/i18n-prerender/.

**Comment by brillout** [comment]
The new hooks are released in `0.2.12`.

Let me know if you run into any (other) issue with i18n.
